### PR TITLE
Multi-select (part 1)

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Implement `Widget.Multiple_Choice` as a multi-select mode for `WidgetSelection`.

https://github.com/enso-org/enso/assets/1047859/49a2ce59-d81a-4bf3-b314-4771b8adf925

### Important Notes

UI:
- Known issue: Some types of changes cause the dropdown to close. This will be addressed in a followup PR.
- Also implements change from #9354: Dropdown selected items are no longer indented.
- The in-node view of the selected elements is a WidgetVector; it allows reordering elements, or remove-by-drag as in other vector-literals. If its add-item button is pressed, the dropdown will be opened to select an item. Any data that doesn't fit the multi-select paradigm (e.g. values outside the allowed set; repeated values) is displayed and can be manipulated in the vector view.

Libs:
- `Widget.Multiple_Choice` can be constructed by the `make_column_name_multi_selector` helper; it can be used in place of the current `make_column_name_vector_selector` when it is expected that each value will be included in the vector at most once.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by jdunkerley — APPROVED** [review]
Nice - approving the Enso side.

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ export interface WidgetEditInteraction extends Interaction {
   start?(origin: PortId): void
   edit?(origin: PortId, value: Ast.Owned | string): void
   end?(origin: PortId): void
+  addItem?(): boolean
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts:23]: The WidgetEditHandler was meant to be for general widget usage; but here I don't know what is this callback exactly. Why it does not provide any info what item was added? What case it should handle? Why this one could stop propagation by returning true? `edit` for example does not allow for that, because every widget could be potentially interested in some child widget edit (it's more like information callback, not one expecting any specific action which parent should handle).

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue [hunk 2]
```diff
@@ -218,44 +199,78 @@ const dropDownInteraction = WidgetEditHandler.New(props.input, {
   end: () => {
     dropdownVisible.value = false
   },
+  addItem: () => {
+    dropdownVisible.value = true
+    editedValue.value = undefined
+    return true
+  },
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue:206]: I don't understand what case we want to handle here. Why we set dropdownVisible to true? Anyone hid it? Why we reset editedValue? If we are multiselect, it is ignored anyway, isn't it?

File: app/gui2/src/components/GraphEditor/widgets/WidgetVector.vue [hunk 3]
```diff
@@ -4,24 +4,26 @@ import ListWidget from '@/components/widgets/ListWidget.vue'
...
 
-const defaultItem = computed(() => {
-  if (props.input.dynamicConfig?.kind === 'Vector_Editor') {
-    return Ast.parse(props.input.dynamicConfig.item_default)
-  } else {
-    return Ast.Wildcard.new(MutableModule.Transient())
-  }
-})
+const itemConfig = computed(() => dynamicConfig.value?.item)
+const defaultItem = computed(() => dynamicConfig.value?.default ?? DEFAULT_ITEM.value)
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/widgets/WidgetVector.vue]: Why this change? Before, we parsed default value from config only when necessary (when adding new item), now we do this on every display (as we need to compute item config to display items).

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue (hunk 2 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue:206]: Oh, I think I know now: the click handler hides the drop down (not canceling the operation, which is my mistake I think) when clicking outside the drop down, even if it's the widget (or any child widget). But we want to any change in child widget keep the drop down open.

So this solution is a bit hacky, and won't work anyway for reordering items, I guess?

When I was implementing the drop down filtering, the drop down indeed was not hidden when clicked inside entire widget (not only `dropdownElement`). It would work well in this case, but I had one problem with it: the default click (if propagated up to WidgetDropDown) toggles the widget visibility, and the dropdow is hidden by _canceling_ interaction, what in turn rejected any user changes in the user's text field, what gave a bad UX.

I'm not sure how to fix the last problem; perhaps toggling drop-down off should end interaction and commit whatever we have in `editedValue` itself.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue (hunk 2 above)
Comment (kazcw) [reply app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue:206]: This isn't in solution to that problem. Sorry, this API really needs documentation--I hadn't written it yet because the refactor to support resumable interactions will affect this code.

The `addItem` interface allows a widget to provide a handler for the add-item intent of a child widget. The vector widget emits it when its plus button is clicked, and the selection widget handles it by opening its dropdown to choose an item. If the parent doesn't handle the event, the vector widget proceeds with its default behavior of inserting a default item.

I hadn't noted the issue you describe as a defect, but now that you bring it up I see that we should change it.
The resumable-interactions refactor will also make that easy to fix.

Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue:206]: Ok, I see. I'm ok with it if documentation will be added.

I already introduced idea of "accepting current edit" in drop downs: https://github.com/enso-org/enso/pull/9536/files#diff-801a0f40269674dc4f01de6d16a7afae776d144bd0ff347c4a0cb0faba5bb69aR207-R211 it could be used in your PR if you find any problems with clicks.

**Review by farmaazon — APPROVED** [review]
